framework/src/araui/core/ui_core.c

```diff
@@ -352,26 +374,45 @@ static void _ui_redraw(uint32_t dt)
 static void _ui_update_redraw_list_recur(ui_widget_body_t *widget)
 {
 	int iter;
-	static bool update_flag = false;
+	ui_mat3_t parent_mat;
+	ui_widget_body_t *curr_widget;
 	ui_widget_body_t *child;
+	bool update_flag;
 
-	if ((widget->update_flag || update_flag) && widget->parent) {
-		// Update previous area
-		ui_window_add_redraw_list(widget->global_rect);
+	ui_widget_queue_init();
+	ui_widget_queue_enqueue(widget);
 
-		ui_renderer_translate(&widget->parent->trans_mat, &widget->trans_mat, (float)widget->local_rect.x, (float)widget->local_rect.y);
-		ui_renderer_rotate(&widget->trans_mat, widget->degree);
-		ui_renderer_scale(&widget->trans_mat, widget->scale_x, widget->scale_y);
+	while (!ui_widget_is_queue_empty()) {
+		curr_widget = ui_widget_queue_dequeue();
+		if (!curr_widget) {
+			continue;
+		}
 
-		// Update new area
-		ui_widget_update_global_rect(widget);
-		ui_window_add_redraw_list(widget->global_rect);
-		widget->update_flag = false;
-		update_flag = true;
-	}
+		if (curr_widget->update_flag || update_flag) {
+			if (curr_widget->parent) {
+				parent_mat = curr_widget->parent->trans_mat;
+			} else {
+				parent_mat = ui_mat3_identity();
+			}
 
-	vec_foreach(&widget->children, child, iter) {
-		_ui_update_redraw_list_recur(child);
+			// Update previous area
+			ui_window_add_redraw_list(curr_widget->global_rect);
```

Comment:
you are modifying your view step by step but you do not check result of each step.
It is very dangerous code if there's no internal checking logic.